test/widgets/actions_test.dart

```diff
@@ -1,341 +1,510 @@
+                },
+                child: const Text('Copy')))))));
+        await tester.pump();
+        await tester.tap(find.text('Copy'));
+        await tester.pump(); // copy
+        await tester.pump(Duration.zero); // await platform info (awkwardly async)
+      }
+
+      Future<void> checkSnackBar(WidgetTester tester, {required bool expected}) async {
+        if (!expected) {
+          check(tester.widgetList(find.byType(SnackBar))).isEmpty();
+          return;
+        }
+        final snackBar = tester.widget<SnackBar>(find.byType(SnackBar));
+        check(snackBar.behavior).equals(SnackBarBehavior.floating);
+        tester.widget(find.descendant(matchRoot: true,
+          of: find.byWidget(snackBar.content), matching: find.text('Text copied')));
+      }
+
+      Future<void> checkClipboardText(String expected) async {
+        check(await Clipboard.getData('text/plain')).isNotNull().text.equals(expected);
+      }
+
+      testWidgets('iOS', (tester) async {
+        testBinding.deviceInfoResult = const IosDeviceInfo(systemVersion: '16.0');
+        await call(tester, text: 'asdf');
+        await checkClipboardText('asdf');
+        await checkSnackBar(tester, expected: true);
+      });
+
+      testWidgets('Android', (tester) async {
+        testBinding.deviceInfoResult = const AndroidDeviceInfo(sdkInt: 33, release: '13');
+        await call(tester, text: 'asdf');
+        await checkClipboardText('asdf');
+        await checkSnackBar(tester, expected: false);
+      });
+
+      testWidgets('Android <13', (tester) async {
+        testBinding.deviceInfoResult = const AndroidDeviceInfo(sdkInt: 32, release: '12');
+        await call(tester, text: 'asdf');
+        await checkClipboardText('asdf');
+        await checkSnackBar(tester, expected: true);
+      });
     });
 
-    testWidgets('catch-all api errors', (tester) async {
-      await prepare(tester);
-      connection.prepare(httpException: http.ClientException('Oops'));
-      final didPass = invokeUpdateMessageFlagsStartingFromAnchor();
-      await tester.pump(Duration.zero);
-      checkErrorDialog(tester,
-        expectedTitle: onFailedTitle,
-        expectedMessage: 'NetworkException: Oops (ClientException: Oops)');
-      check(await didPass).isFalse();
+    group('launchUrl', () {
+      Future<void> call(WidgetTester tester, {required Uri url}) async {
+        await tester.pumpWidget(TestZulipApp(
+          child: Builder(builder: (context) => Center(
+            child: ElevatedButton(
+              onPressed: () async {
+                await PlatformActions.launchUrl(context, url);
+              },
+              child: const Text('link'))))));
+        await tester.pump();
+        await tester.tap(find.text('link'));
+        await tester.pump(Duration.zero);
+      }
+
+      final httpUrl = Uri.parse('http://chat.zulip.org');
+      final nonHttpUrl = Uri.parse('mailto:chat@zulip.org');
```

Comment:
nit:
```suggestion
      final httpUrl = Uri.parse('https://chat.example');
      final nonHttpUrl = Uri.parse('mailto:chat@example');
```

* use reserved example domains (RFC 6761; https://en.wikipedia.org/wiki/Special-use_domain_name)
* use `https:` rather than `http:`, since `https:` should be the norm and cleartext `http:` rare